ZFSin/zfs/module/zfs/zfs_windows_zvol_scsi.c

```diff
@@ -792,28 +808,42 @@ wzvol_WkRtn(__in PVOID pWkParms)                          // Parm list pointer.
 		goto Done;
 	}
 
-
 	if (sectorOffset >= zv->zv_volsize) {      // Starting sector beyond the bounds?
 		dprintf("%s: invalid starting sector: %d\n", __func__, startingSector);
 		status = SRB_STATUS_INVALID_REQUEST;
 		goto Done;
 	}
 
```

Comment:
Global comment about the entire processing for the reset handling: 
the way you have done it runs the risk of never (or delaying for a long time) clearing zv_request_cancellation if you receive a lot of data because each I/O is queued in a work item.  zv_requests_processing_count could stay above 0 for a long time.  Normally the initiator stops sending I/O when triggering a RESET but remember that the RESET will be replied right away with a completed status so the initiator could flood the LUN with I/Os after that and this could prevent your count from going to 0 , which in practice could leave you in the cancelling state virtually "forever" or too long at least.  also, after the initiator gets back form the RESEt command it sent it expects further I/Os to *not* be cancelled but in your scheme those will be and that is not correct. 
Ideally ZFS should not use work items to queue the I/Os or they should also be queued in a control queue so the RESET processing would mark them all as cancelled and the srb status would be set later accordingly.  That would be too much change right now.  
At least, if you notice I/Os should be cancelled you should not do the processing *after* you send the read/write I/O but earlier in this routine.  There's no point sending the I/O in zvol_read_win or zvol_write_win if you have the cancel bit set.   I would add that if the cancel bit is set in the zv we should not even queue the I/O in the work item at all and fail it right away with SRB status BUSY then.  No need to queue them.
And then lastly the way you coded the wzvol_reset_notification will probably not work if you get a SRB_FUNCTION_RESET_DEVICE because in that case the LUN value in the Srb is not set or should ignored.  All LUNs in the target should set the zv_request_cancellation bit.  In the current implementation of ZFS there is only lun 0 for each target so it will work somewhat but the search should then specify 0 in that case, not the Lun number in the Srb for a reset device.  The logic will also break if one day someone really uses the lun numbers under the target so I would right away change that wzvol_reset_notification routine for RESET DEVICE to locate all zv's under the same target id (no matter the Lun number) and set the cancel bit for all of those.  It's going to be one today but the logic is future-proof.
In summary after the above I would say that look at the condition that will make you exit the zv_request_cancellation state to be sure you have a deterministic way of getting out of it.  Only I/Os received before the RESET command should be cancelled, not the ones after.